rust/kernel/sync/arc.rs

```diff
@@ -0,0 +1,166 @@
+// SPDX-License-Identifier: GPL-2.0
+
+//! A reference-counted pointer.
+//!
+//! This module implements a way for users to create reference-counted objects and pointers to
+//! them. Such a pointer automatically increments and decrements the count, and drops the
+//! underlying object when it reaches zero.
+//!
+//! It is different from the standard library's [`Arc`] in two ways: it does not support weak
+//! references, which allows it to be smaller -- a single pointer-sized integer; it allows users to
+//! safely increment the reference count from a single reference to the underlying object.
+//!
+//! [`Arc`]: https://doc.rust-lang.org/std/sync/struct.Arc.html
+
+use crate::KernelResult;
+use alloc::boxed::Box;
+use core::{
+    mem::ManuallyDrop,
+    ops::Deref,
+    ptr::NonNull,
+    sync::atomic::{AtomicUsize, Ordering},
+};
+
+/// A reference-counted pointer to an instance of `T`.
+///
+/// The reference count is incremented when new instances of [`Ref`] are created, and decremented
+/// when they are dropped. When the count reaches zero, the underlying `T` is also dropped.
+///
+/// # Invariants
+///
+/// The value stored in [`RefCounted::get_count`] corresponds to the number of instances of [`Ref`]
+/// that point to that instance of `T`.
+pub struct Ref<T: RefCounted + ?Sized> {
+    ptr: NonNull<T>,
+}
+
+impl<T: RefCounted> Ref<T> {
+    /// Constructs a new reference counted instance of `T`.
+    pub fn try_new(contents: T) -> KernelResult<Self> {
+        let boxed = Box::try_new(contents)?;
+        boxed.get_count().count.store(1, Ordering::Release);
+        let ptr = NonNull::from(boxed.deref());
+        Box::into_raw(boxed);
```

Comment:
Can this just be `NonNull::new(boxed.into_raw())`?

Reply:
`NonNull::new` may fail (if its argument is `null`), in which case it returns `None`. `NonNull::from` OTOH never fails because its argument cannot be `null` -- by doing it the way I did I have one less failure path.

Reply:
Ok. I think this maybe makes more sense as `mem::forget(boxed)` then? We're not actually using the `into_raw()` ptr, which momentarily threw me.

Reply:
Now that I think about it, it is possible that some future version of `Box::into_raw` will return something else other than `boxed.deref()`, so I think it's better to use your original suggestion, even if it implies an extra failure path (as it should never happen, it's possible the optimiser remove it from the final binary).

Reply:
Oh, look at this: https://github.com/rust-lang/rust/issues/47336

So the way to go is `NonNull::from(Box::leak(boxed))` -- no failure path and still compatible with `Box::from_raw`!